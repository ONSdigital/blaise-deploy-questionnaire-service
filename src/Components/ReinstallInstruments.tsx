import React, {Fragment, ReactElement, useEffect, useState} from "react";
import {Link, Redirect} from "react-router-dom";
import {getAllInstrumentsInBucket} from "../utilities/http";
import {ErrorBoundary, ONSButton, ONSLoadingPanel, ONSPanel} from "blaise-design-system-react-components";
import {Instrument} from "../../Interfaces";
import {verifyAndInstallInstrument} from "../utilities/processes";

interface Props {
    installedInstruments: Instrument[];
    listLoading: boolean;
}

function ReinstallInstruments({installedInstruments, listLoading}: Props): ReactElement {
    const [instrumentList, setInstrumentList] = useState<string[]>([]);
    const [listError, setListError] = useState<string>("Loading ...");
    const [redirect, setRedirect] = useState<boolean>(false);
    const [uploadStatus, setUploadStatus] = useState<string>("");
    const [instrumentName, setInstrumentName] = useState<string>("");
    const [instrumentToInstall, setInstrumentToInstall] = useState<string>("");
    const [loading, setLoading] = useState<boolean>(true);
    const [installing, setInstalling] = useState<boolean>(false);

    useEffect(() => {
        getInstruments().then();
    }, [installedInstruments]);

    function getInstalledInstrumentList() {
        const list: string[] = [];
        installedInstruments.map(({name}) => {
            list.push(`${name}.bpkg`);
        });
        return list;
    }

    async function getInstruments() {
        setLoading(true);
        console.log("getInstruments");
        const list: string[] = [];
        setListError("");

        const [success, bucketInstrumentList] = await getAllInstrumentsInBucket();

        if (!success) {
            setListError("Unable to load questionnaires.");
            setLoading(false);
            return;
        }

        const installedInstrumentList = getInstalledInstrumentList();

        bucketInstrumentList.map((instrument) => {
            if (!installedInstrumentList.includes(instrument)) {
                list.push(instrument);
            }
        });

        if (list.length === 0) {
            setListError("No compatible previously installed questionnaires found.");
        }

        setInstrumentList(list);
        setLoading(false);
    }

    async function installInstrumentFromBucket() {
        setInstalling(true);
        setInstrumentName(instrumentToInstall.replace(/\.[a-zA-Z]*$/, ""));

        const [installed, message] = await verifyAndInstallInstrument(instrumentToInstall);
        if (!installed) {
            setUploadStatus(message);
        }

        setRedirect(true);
    }


    function DisplayInstrumentsToInstallList(): ReactElement {
        return (
            <div className={"elementToFadeIn"}>
                <ErrorBoundary errorMessageText={"Failed to Unable to load questionnaires."}>
                    <form>
                        {
                            listError !== "" ?
                                <ONSPanel>{listError}</ONSPanel>
                                :
                                <>
                                    <fieldset className="fieldset">
                                        <legend className="fieldset__legend">
                                            Select a questionnaire to install
                                        </legend>
                                        <div className="radios__items">
                                            {
                                                instrumentList.map((item: string) => {
                                                    return (
                                                        <Fragment key={item}>
                                                            <p className="radios__item">
                                                            <span className="radio">
                                                                <input
                                                                    type="radio"
                                                                    id={`install-${item}`}
                                                                    className="radio__input js-radio "
                                                                    value={item}
                                                                    name="select-survey"
                                                                    aria-label="No"
                                                                    onChange={() => setInstrumentToInstall(item)}
                                                                />
                                                                <label className="radio__label "
                                                                       htmlFor={`install-${item}`}>
                                                                    {item}
                                                                </label>
                                                            </span>
                                                            </p>
                                                            <br/>
                                                        </Fragment>
                                                    );
                                                })
                                            }
                                        </div>
                                    </fieldset>
                                    <br/>

                                    <ONSButton
                                        label={"Install selected questionnaire"}
                                        primary={true}
                                        loading={installing}
                                        id="confirm-install"
                                        onClick={() => installInstrumentFromBucket()}/>
                                </>
                        }
                    </form>
                </ErrorBoundary>
            </div>
        );
    }

    return (
        <>
            {
                redirect && <Redirect
                    to={{
                        pathname: "/UploadSummary",
                        state: {questionnaireName: instrumentName, status: uploadStatus}
                    }}/>
            }
            <p>
                <Link to={"/"}>Previous</Link>
            </p>
            <h1>Reinstall questionnaire</h1>
            {
                (listLoading || loading) ?
                    <ONSLoadingPanel/>
                    :
                    DisplayInstrumentsToInstallList()
            }
        </>
    );
}

export default ReinstallInstruments;
