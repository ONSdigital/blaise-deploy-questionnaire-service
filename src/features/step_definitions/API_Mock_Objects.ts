import { Survey } from "../../../Interfaces";

export const survey_list_with_OPN_and_LMS_with_one_active_instrument_each: Survey[] = [
    {
        survey: "OPN",
        instruments: [
            {
                activeToday: true,
                fieldPeriod: "July 2020",
                expired: false,
                installDate: "2020-12-11T11:53:55.5612856+00:00",
                link: "https://external-web-url/OPN2007T?LayoutSet=CATI-Interviewer_Large",
                name: "OPN2007T",
                serverParkName: "LocalDevelopment",
                surveyTLA: "OPN",
                surveyDays: []
            }
        ]
    },
    {
        survey: "LMS",
        instruments: [
            {
                activeToday: true,
                fieldPeriod: "Unknown",
                expired: false,
                installDate: "2020-12-11T11:53:55.5612856+00:00",
                link: "https://external-web-url/LMSSurvey01?LayoutSet=CATI-Interviewer_Large",
                name: "LMSSurvey01",
                serverParkName: "LocalDevelopment",
                surveyTLA: "LMS",
                surveyDays: []
            }
        ]
    }
];

export const survey_list_with_OPN_with_one_active_instrument: Survey[] = [
    {
        survey: "OPN",
        instruments: [
            {
                activeToday: true,
                fieldPeriod: "July 2020",
                expired: false,
                installDate: "2020-12-11T11:53:55.5612856+00:00",
                link: "https://external-web-url/OPN2007T?LayoutSet=CATI-Interviewer_Large",
                name: "OPN2007T",
                serverParkName: "LocalDevelopment",
                surveyTLA: "OPN",
                surveyDays: []
            }
        ]
    }
];

export const survey_list_with_OPN_with_two_active_instruments: Survey[] = [
    {
        survey: "OPN",
        instruments: [
            {
                activeToday: true,
                fieldPeriod: "July 2020",
                expired: false,
                installDate: "2020-12-10T11:53:55.5612856+00:00",
                link: "https://external-web-url/OPN2007T?LayoutSet=CATI-Interviewer_Large",
                name: "OPN2007T",
                serverParkName: "LocalDevelopment",
                surveyTLA: "OPN",
                surveyDays: []
            },
            {
                activeToday: true,
                fieldPeriod: "April 2020",
                expired: false,
                installDate: "2020-12-11T11:53:55.5612856+00:00",
                link: "https://external-web-url/OPN2004A?LayoutSet=CATI-Interviewer_Large",
                name: "OPN2004A",
                serverParkName: "LocalDevelopment",
                surveyTLA: "OPN",
                surveyDays: []
            }
        ]
    }
];

export const survey_list_with_OPN_with_three_active_instruments: Survey[] = [
    {
        survey: "OPN",
        instruments: [
            {
                activeToday: true,
                fieldPeriod: "July 2020",
                expired: false,
                installDate: "2020-12-10T11:53:55.5612856+00:00",
                link: "https://external-web-url/OPN2007T?LayoutSet=CATI-Interviewer_Large",
                name: "OPN2007T",
                serverParkName: "LocalDevelopment",
                surveyTLA: "OPN",
                surveyDays: []
            },
            {
                activeToday: true,
                fieldPeriod: "January 2021",
                expired: false,
                installDate: "2020-12-12T11:53:55.5612856+00:00",
                link: "https://external-web-url/OPN2101A?LayoutSet=CATI-Interviewer_Large",
                name: "OPN2101A",
                serverParkName: "LocalDevelopment",
                surveyTLA: "OPN",
                surveyDays: []
            },
            {
                activeToday: true,
                fieldPeriod: "April 2020",
                expired: false,
                installDate: "2020-12-11T11:53:55.5612856+00:00",
                link: "https://external-web-url/OPN2004A?LayoutSet=CATI-Interviewer_Large",
                name: "OPN2004A",
                serverParkName: "LocalDevelopment",
                surveyTLA: "OPN",
                surveyDays: []
            },
        ]
    }
];
