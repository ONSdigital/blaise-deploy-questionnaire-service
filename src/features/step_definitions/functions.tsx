import {createMemoryHistory} from "history";
import {act, fireEvent, render, screen} from "@testing-library/react";
import {Router} from "react-router";
import App from "../../App";
import flushPromises from "../../tests/utils";
import React from "react";

/*
* Renders the App in a Mock Router, then navigates the the 'Deploy a questionnaire' page
* and then selects a mock OPN2004A.bpkg in the File select Input field.
*  */
export default async function navigateToDeployPageAndSelectFile() {
    const history = createMemoryHistory();
    render(
        <Router history={history}>
            <App/>
        </Router>
    );
    await act(async () => {
        await flushPromises();
    });

    await fireEvent.click(screen.getByText(/Deploy a questionnaire/i));

    const inputEl = screen.getByLabelText(/Select survey package/i);

    const file = new File(["(⌐□_□)"], "OPN2004A.bpkg", {
        type: "bpkg"
    });

    Object.defineProperty(inputEl, "files", {
        value: [file]
    });

    fireEvent.change(inputEl);
}

export async function navigatePastSettingTOStartDateAndStartDeployment() {
    await act(async () => {
        await flushPromises();
    });
    await fireEvent.click(screen.getByText(/No start date/i));
    await fireEvent.click(screen.getByText(/Continue/));

    await act(async () => {
        await flushPromises();
    });

    await fireEvent.click(screen.getByText(/Deploy questionnaire/));

    await act(async () => {
        await flushPromises();
    });
}

export function mock_fetch_requests(mock_server_responses: any) {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    global.fetch = jest.fn((url: string) => mock_server_responses(url));
}
