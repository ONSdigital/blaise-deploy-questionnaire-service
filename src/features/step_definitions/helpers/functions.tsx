import { createMemoryHistory } from "history";
import { act, render, screen } from "@testing-library/react";
import { Router } from "react-router";
import App from "../../../app";
import flushPromises from "../../../tests/utils";
import React from "react";
import userEvent from "@testing-library/user-event";

/*
* Renders the App in a Mock Router, then navigates the the 'Deploy a questionnaire' page
* and then selects a mock OPN2004A.bpkg in the File select Input field.
*  */
export async function navigateToDeployPageAndSelectFile(questionnaire = "OPN2004A"): Promise<void> {
    const history = createMemoryHistory();
    render(
        <Router history={history}>
            <App />
        </Router>
    );
    await act(async () => {
        await flushPromises();
    });

    userEvent.click(screen.getByText(/Deploy a questionnaire/i));

    const input = screen.getByLabelText(/Select survey package/i);

    const file = new File(["(⌐□_□)"], `${questionnaire}.bpkg`, { type: "application/zip" });

    userEvent.upload(input, file);
}

export default async function navigateToDeployPageAndSelectFileAndContinue(questionnaire = "OPN2004A"): Promise<void> {
    await navigateToDeployPageAndSelectFile(questionnaire)
    await clickContinue();
}

export async function navigatePastSettingTOStartDateAndDeployQuestionnaire(): Promise<void> {
    await selectNoTOStartDateAndContinue();
    await clickContinue();
    await clickDeployQuestionnaire();
}

export async function navigateToSetTMReleaseDatePageAndContinue(questionnaire: string): Promise<void> {
    await navigateToDeployPageAndSelectFile(questionnaire);
    await clickContinue();

    // TO start date page
    await selectNoTOStartDateAndContinue();
    await clickContinue();
}

export async function selectNoTOStartDateAndContinue(): Promise<void> {
    userEvent.click(screen.getByText(/No start date/i));
}

export async function selectNoTMReleaseDateAndContinue(): Promise<void> {
    userEvent.click(screen.getByText(/No release date/i));
    userEvent.click(screen.getByText(/Continue/));
    await act(async () => {
        await flushPromises();
    });
}

export async function clickDeployQuestionnaire(): Promise<void> {
    userEvent.click(screen.getByText(/Deploy questionnaire/));
    await act(async () => {
        await flushPromises();
    });
}

export async function clickContinue(): Promise<void> {
    userEvent.click(screen.getByText(/Continue/));
    await act(async () => {
        await flushPromises();
    });
}

export function formatDateString(date: string): string {
    const splitDate = date.split("/");
    return `${splitDate[2]}-${splitDate[1]}-${splitDate[0]}`;
}
