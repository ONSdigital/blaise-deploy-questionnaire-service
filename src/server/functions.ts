function isNumber(n: any) {
    return !isNaN(parseFloat(n)) && isFinite(n);
}

function fieldPeriodFromInstrument(instrument_name: string): string {
    const month_number_str: string = instrument_name.substring(5, 7);
    let month_number_int = -1;
    let month = "Unknown";

    const monthNames = ["January", "February", "March", "April", "May", "June",
        "July", "August", "September", "October", "November", "December"
    ];

    if (isNumber(month_number_str)) {
        month_number_int = parseInt(month_number_str) - 1;
    } else {
        throw "Month was not a integer";
    }

    if (month_number_int >= 0 && month_number_int < 12) {
        month = monthNames[month_number_int];
    } else {
        throw "Month was dot between 1 and 12";
    }

    return month + " 20" + instrument_name.substring(3, 5);
}

export function fieldPeriodToText(instrument_name: string): string {
    try {
        return fieldPeriodFromInstrument(instrument_name);
    } catch (error: unknown) {
        console.error("Error getting field period");
        return "Field period unknown";
    }
}
