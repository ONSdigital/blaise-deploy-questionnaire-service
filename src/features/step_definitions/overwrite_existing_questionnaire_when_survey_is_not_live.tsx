// React
import React from "react";
// Test modules
import {defineFeature, loadFeature} from "jest-cucumber";
import {cleanup, fireEvent, screen, waitFor} from "@testing-library/react";
import "@testing-library/jest-dom";
// Mock elements
import {survey_list} from "./API_Mock_Objects";
import navigateToDeployPageAndSelectFile, {mock_fetch_requests} from "./functions";


// Mock the Uploader.js module
jest.mock("../../uploader");


// Load in feature details from .feature file
const feature = loadFeature(
    "./src/features/overwrite_existing_questionnaire_when_survey_is_not_live.feature",
    {tagFilter: "not @server and not @integration"}
);

const mock_server_responses = (url: string) => {
    console.log(url);
    if (url.includes("bucket")) {
        return Promise.resolve({
            status: 200,
            json: () => Promise.resolve({name: "OPN2004A.bpkg"}),
        });
    } else if (url.includes("/api/install")) {
        return Promise.resolve({
            status: 500,
            json: () => Promise.resolve({}),
        });
    } else {
        return Promise.resolve({
            status: 200,
            json: () => Promise.resolve(survey_list),
        });
    }
};

defineFeature(feature, test => {
    afterEach(() => {
        jest.clearAllMocks();
        cleanup();
        jest.resetModules();
    });

    beforeEach(() => {
        cleanup();
    });

});
