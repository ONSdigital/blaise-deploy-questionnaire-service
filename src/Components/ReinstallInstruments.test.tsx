import React from "react";
import {render, waitFor, cleanup, screen, fireEvent} from "@testing-library/react";
import "@testing-library/jest-dom";
import flushPromises, {mock_server_request_Return_JSON} from "../tests/utils";
import {act} from "react-dom/test-utils";
import {createMemoryHistory} from "history";
import {Router} from "react-router";
import {Instrument} from "../../Interfaces";
import ReinstallInstruments from "./ReinstallInstruments";
import {mock_fetch_requests} from "../features/step_definitions/functions";


const instrumentList: Instrument[] = [{
    name: "OPN2101A",
    serverParkName: "gusty",
    installDate: "2021-01-15T14:41:29.4399898+00:00",
    status: "Active",
    dataRecordCount: 0,
    hasData: false,
    active: false,
    fieldPeriod: "January 2021"
}];

const bucketInstrument: string[] = ["OPN2101A.bpkg", "OPN2004A.bpkg", "LMS2101_BK2.bpkg"];

describe("Reinstall instruments list", () => {

    beforeAll(() => {
        mock_server_request_Return_JSON(200, bucketInstrument);
    });

    it("view Blaise Status page matches Snapshot", async () => {
        const history = createMemoryHistory();
        const wrapper = render(
            <Router history={history}>
                <ReinstallInstruments installedInstruments={instrumentList}  listLoading={false}/>
            </Router>
        );

        await act(async () => {
            await flushPromises();
        });

        await waitFor(() => {
            expect(wrapper).toMatchSnapshot();
        });
    });

    it("should render correctly", async () => {
        const history = createMemoryHistory();
        render(
            <Router history={history}>
                <ReinstallInstruments installedInstruments={instrumentList} listLoading={false}/>
            </Router>
        );

        await act(async () => {
            await flushPromises();
            await flushPromises();
        });

        await waitFor(() => {
            expect(screen.getByText(/Reinstall questionnaire/i)).toBeDefined();
            expect(screen.getByText(/OPN2004A/i)).toBeDefined();
            expect(screen.getByText(/LMS2101_BK2/i)).toBeDefined();
            expect(screen.queryByText(/OPN2101A/i)).not.toBeInTheDocument();
        });
    });

    it("should render a message if all items in bucket are already installed", async () => {
        const instrumentList: any[] = [{name: "OPN2101A"}, {name: "OPN2004A"}, {name: "LMS2101_BK2"}];
        const history = createMemoryHistory();
        render(
            <Router history={history}>
                <ReinstallInstruments installedInstruments={instrumentList} listLoading={false}/>
            </Router>
        );

        await act(async () => {
            await flushPromises();
        });

        await waitFor(() => {
            expect(screen.getByText(/Reinstall questionnaire/i)).toBeDefined();
            expect(screen.getByText(/No compatible previously installed questionnaires found./i)).toBeDefined();
        });
    });

    afterAll(() => {
        jest.clearAllMocks();
        cleanup();
    });
});

describe("Reinstall an instruments", () => {
    const mock_server_responses = (url: string) => {
        console.log(url);
        if (url.includes("/bucket/files")) {
            return Promise.resolve({
                status: 404,
                json: () => Promise.resolve(bucketInstrument),
            });
        } else if (url.includes("/upload/verify")) {
            return Promise.resolve({
                status: 200,
                json: () => Promise.resolve({name: "OPN2004A.bpkg"}),
            });
        } else if (url.includes("/api/install")) {
            return Promise.resolve({
                status: 201,
                json: () => Promise.resolve(),
            });
        } else {
            return Promise.resolve({
                status: 200,
                json: () => Promise.resolve([]),
            });
        }
    };
    beforeAll(() => {
        mock_fetch_requests(mock_server_responses);
    });

    it("should redirect to the success page after install", async () => {
        const history = createMemoryHistory();
        render(
            <Router history={history}>
                <ReinstallInstruments installedInstruments={instrumentList}  listLoading={false}/>
            </Router>
        );

        await act(async () => {
            await flushPromises();
        });

        await act(async () => {
            fireEvent.click(screen.getByText(/OPN2004A/i));
            await flushPromises();
            fireEvent.click(screen.getByText(/Install selected questionnaire/i));
            await flushPromises();
        });

        await waitFor(() => {
            // Check page has been redirected to summary page
            expect(history.location.pathname).toEqual("/UploadSummary");
            // State should be blank as its successful
            expect(history.location.state).toEqual({questionnaireName: "OPN2004A", status: ""});
        });
    });

    it("should render a message if all items in bucket are already installed", async () => {
        const instrumentList: any[] = [{name: "OPN2101A"}, {name: "OPN2004A"}, {name: "LMS2101_BK2"}];
        const history = createMemoryHistory();
        render(
            <Router history={history}>
                <ReinstallInstruments installedInstruments={instrumentList} listLoading={false}/>
            </Router>
        );

        await act(async () => {
            await flushPromises();
        });

        await waitFor(() => {
            expect(screen.getByText(/Reinstall questionnaire/i)).toBeDefined();
            expect(screen.getByText(/No compatible previously installed questionnaires found./i)).toBeDefined();
        });
    });

    afterAll(() => {
        jest.clearAllMocks();
        cleanup();
    });
});




describe("Given the API returns a 500 status", () => {

    beforeAll(() => {
        mock_server_request_Return_JSON(500, []);
    });

    it("it should render with the error message displayed", async () => {
        const history = createMemoryHistory();
        render(
            <Router history={history}>
                <ReinstallInstruments installedInstruments={instrumentList} listLoading={false}/>
            </Router>
        );


        await waitFor(() => {
            expect(screen.getByText(/Unable to load questionnaires./i)).toBeDefined();
            expect(screen.queryByText(/Loading/i)).not.toBeInTheDocument();
        });

    });

    afterAll(() => {
        jest.clearAllMocks();
        cleanup();
    });
});

describe("Given the API returns an empty list", () => {

    beforeAll(() => {
        mock_server_request_Return_JSON(200, []);
    });

    it("it should render with a message to inform the user in the list", async () => {
        const history = createMemoryHistory();
        render(
            <Router history={history}>
                <ReinstallInstruments installedInstruments={instrumentList} listLoading={false}/>
            </Router>
        );

        await act(async () => {
            await flushPromises();
        });

        await waitFor(() => {
            expect(screen.getByText(/Reinstall questionnaire/i)).toBeDefined();
            expect(screen.getByText(/No compatible previously installed questionnaires found./i)).toBeDefined();
        });
    });

    afterAll(() => {
        jest.clearAllMocks();
        cleanup();
    });
});
