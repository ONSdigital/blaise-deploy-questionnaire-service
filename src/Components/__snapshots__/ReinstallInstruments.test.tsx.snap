// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reinstall instruments list view Blaise Status page matches Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        aria-label="Breadcrumb"
        class="breadcrumb"
      >
        <ol
          class="breadcrumb__items u-fs-s"
        >
          <li
            class="breadcrumb__item"
            id="breadcrumb-0"
          >
            <a
              class="breadcrumb__link"
              href="/"
            >
              Home
            </a>
            <svg
              class="svg-icon"
              focusable="false"
              viewBox="0 0 8 13"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
                transform="translate(-5.02 -1.59)"
              />
            </svg>
          </li>
          <li
            class="breadcrumb__item"
            id="breadcrumb-1"
          >
            <a
              class="breadcrumb__link"
              href="/upload"
            >
              Deploy a questionnaire
            </a>
            <svg
              class="svg-icon"
              focusable="false"
              viewBox="0 0 8 13"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
                transform="translate(-5.02 -1.59)"
              />
            </svg>
          </li>
        </ol>
      </nav>
      <main
        class="page__main u-mt-no"
        id="main-content"
      >
        <h1
          class="u-mb-l"
        >
          Reinstall questionnaire
        </h1>
        <p>
          Reinstall a previously uploaded questionnaire.
          <br />
          This will always deploy the last uploaded version of the questionnaire.
        </p>
        <div
          class="elementToFadeIn"
        >
          <form>
            <fieldset
              class="fieldset"
            >
              <legend
                class="fieldset__legend"
              >
                Select a questionnaire to install
              </legend>
              <div
                class="radios__items"
              >
                <p
                  class="radios__item"
                >
                  <span
                    class="radio"
                  >
                    <input
                      aria-label="No"
                      class="radio__input js-radio "
                      id="install-OPN2004A.bpkg"
                      name="select-survey"
                      type="radio"
                      value="OPN2004A.bpkg"
                    />
                    <label
                      class="radio__label "
                      for="install-OPN2004A.bpkg"
                    >
                      OPN2004A.bpkg
                    </label>
                  </span>
                </p>
                <br />
                <p
                  class="radios__item"
                >
                  <span
                    class="radio"
                  >
                    <input
                      aria-label="No"
                      class="radio__input js-radio "
                      id="install-LMS2101_BK2.bpkg"
                      name="select-survey"
                      type="radio"
                      value="LMS2101_BK2.bpkg"
                    />
                    <label
                      class="radio__label "
                      for="install-LMS2101_BK2.bpkg"
                    >
                      LMS2101_BK2.bpkg
                    </label>
                  </span>
                </p>
                <br />
              </div>
            </fieldset>
            <br />
            <button
              class="btn "
              data-testid="button"
              id="confirm-install"
              type="button"
            >
              <span
                class="btn__inner"
              >
                Install selected questionnaire
              </span>
            </button>
          </form>
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <nav
      aria-label="Breadcrumb"
      class="breadcrumb"
    >
      <ol
        class="breadcrumb__items u-fs-s"
      >
        <li
          class="breadcrumb__item"
          id="breadcrumb-0"
        >
          <a
            class="breadcrumb__link"
            href="/"
          >
            Home
          </a>
          <svg
            class="svg-icon"
            focusable="false"
            viewBox="0 0 8 13"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
              transform="translate(-5.02 -1.59)"
            />
          </svg>
        </li>
        <li
          class="breadcrumb__item"
          id="breadcrumb-1"
        >
          <a
            class="breadcrumb__link"
            href="/upload"
          >
            Deploy a questionnaire
          </a>
          <svg
            class="svg-icon"
            focusable="false"
            viewBox="0 0 8 13"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
              transform="translate(-5.02 -1.59)"
            />
          </svg>
        </li>
      </ol>
    </nav>
    <main
      class="page__main u-mt-no"
      id="main-content"
    >
      <h1
        class="u-mb-l"
      >
        Reinstall questionnaire
      </h1>
      <p>
        Reinstall a previously uploaded questionnaire.
        <br />
        This will always deploy the last uploaded version of the questionnaire.
      </p>
      <div
        class="elementToFadeIn"
      >
        <form>
          <fieldset
            class="fieldset"
          >
            <legend
              class="fieldset__legend"
            >
              Select a questionnaire to install
            </legend>
            <div
              class="radios__items"
            >
              <p
                class="radios__item"
              >
                <span
                  class="radio"
                >
                  <input
                    aria-label="No"
                    class="radio__input js-radio "
                    id="install-OPN2004A.bpkg"
                    name="select-survey"
                    type="radio"
                    value="OPN2004A.bpkg"
                  />
                  <label
                    class="radio__label "
                    for="install-OPN2004A.bpkg"
                  >
                    OPN2004A.bpkg
                  </label>
                </span>
              </p>
              <br />
              <p
                class="radios__item"
              >
                <span
                  class="radio"
                >
                  <input
                    aria-label="No"
                    class="radio__input js-radio "
                    id="install-LMS2101_BK2.bpkg"
                    name="select-survey"
                    type="radio"
                    value="LMS2101_BK2.bpkg"
                  />
                  <label
                    class="radio__label "
                    for="install-LMS2101_BK2.bpkg"
                  >
                    LMS2101_BK2.bpkg
                  </label>
                </span>
              </p>
              <br />
            </div>
          </fieldset>
          <br />
          <button
            class="btn "
            data-testid="button"
            id="confirm-install"
            type="button"
          >
            <span
              class="btn__inner"
            >
              Install selected questionnaire
            </span>
          </button>
        </form>
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
